docs/architecture/adr-006-02-client-refactor.md

```diff
@@ -0,0 +1,170 @@
+To work around the lack of support for non-zero height upgrades, an abstract height type was created along with an upgrade mechanism. 
+This type would indicate the revision number (the number of times the chain-id has been changed) and revision height (the current height of the blockchain). 
+
+Refs: 
+- Issue [#439](https://github.com/cosmos/ibc/issues/439) on IBC specification repository.
+- Specification changes in [#447](https://github.com/cosmos/ibc/pull/447)
+- Implementation changes for the abstract height type, [SDK#7211](https://github.com/cosmos/cosmos-sdk/pull/7211)
+
+### Tendermint requires misbehaviour detection during updates
+
+The initial release of the IBC module and the 07-tendermint light client implementation did not support misbehaviour detection during update nor did it prevent overwriting of previous updates.
+Despite the fact that the we designed the `ClientState` interface and developed the 07-tendermint client, we failed to detect even a duplicate update that constituted misbehaviour and thus should freeze the client. 
+This was fixed in pr [#141](https://github.com/cosmos/ibc-go/pull/141) which required light client implementations to be aware that they must handle duplicate updates and misbehaviour detection.
+Misbehaviour detection during updates is not applicable to the solomachine nor localhost.
+It is also not obvious that `CheckHeaderAndUpdateState` should be performing this functionality. 
+
+### Localhost requires access to the entire client store
+
+The localhost has been broken since the initial version of the IBC module. 
+The localhost tried to be developed underneath the 02-client interfaces without special exception, but this proved to be impossible. 
+The issues was outlined in [#27](https://github.com/cosmos/ibc-go/issues/27) and a further discussed in the attempted ADR in [#75](https://github.com/cosmos/ibc-go/pull/75).
+Unlike all other clients, the localhost requires access to the entire IBC store and not just the prefixed client store.
+
+### Solomachine doesn't set consensus states 
+
+The 06-solomachine does not set the consensus states within the prefixed client store.
+It has a single consensus state that is stored within the client state. 
+This causes setting of the consensus state at the 02-client level to be unnecessary storage.
+It also causes timeouts to fail with solo machines. 
+Previously, the timeout logic within IBC would obtain the consensus state at the height a timeout is being proved.
+This is problematic for the solo machine as no consensus state is set. 
+See issue [#562](https://github.com/cosmos/ibc/issues/562) on the IBC specification repo. 
+
+### New clients may want to do batch updates
+
+New light clients may not function in a similar fashion to 06-solomachine and 07-tendermint. 
+They may require setting many consensus states in a single update. 
+As @seunlanlege [states](https://github.com/cosmos/ibc-go/issues/284#issuecomment-1005583679):
+```
```

Comment:
Instead of formatting this with triple backtick maybe it's better to use `>` so that lines break and it's easier to read (i.e. you don't need to scroll to the right).

Reply:
fixed